validator/src/service/machine.ts

```diff
@@ -0,0 +1,499 @@
+		protocol: ShieldnetProtocol;
+		keyGenClient: KeyGenClient;
+		signingClient: SigningClient;
+		verificationEngine: VerificationEngine;
+		initialEpoch?: bigint;
+		logger?: (msg: unknown) => void;
+		blocksPerEpoch?: bigint;
+	}) {
+		this.#participants = participants;
+		this.#protocol = protocol;
+		this.#keyGenClient = keyGenClient;
+		this.#signingClient = signingClient;
+		this.#verificationEngine = verificationEngine;
+		this.#activeEpoch = initialEpoch ?? 0n;
+		this.#logger = logger;
+		this.#blocksPerEpoch = blocksPerEpoch ?? BLOCKS_PER_EPOCH;
+	}
+
+	async progressToBlock(block: bigint) {
+		if (block < this.#lastProcessedBlock) {
+			throw Error(`There is no going back!`);
+		}
+		// Check if we are already up to date
+		if (block === this.#lastProcessedBlock) return;
+		this.#lastProcessedBlock = block;
+
+		await this.checkGenesis();
+		await this.checkEpochRollover(block);
+		await this.checkAvailableNonces();
+
+		// TODO:
+		// Check keyGen timeouts
+		// Check signing timeouts
+	}
+
+	async processBlockEvent(
+		block: bigint,
+		index: number,
+		eventName: string,
+		eventArgs: unknown,
+	) {
+		if (
+			block < this.#lastProcessedBlock ||
+			(block === this.#lastProcessedBlock && index <= this.#lastProcessedIndex)
+		) {
+			throw Error(
+				`Invalid block number (${block}) and index ${index} (currentyl at block ${this.#lastProcessedBlock} and index ${this.#lastProcessedIndex})`,
+			);
+		}
+		this.#lastProcessedIndex = index;
+		await this.progressToBlock(block);
+		await this.handleEvent(eventName, eventArgs);
+		// Check after every event if we could do a epoch rollover
+		await this.checkEpochRollover(block);
+	}
+
+	private async handleEvent(eventName: string, eventArgs: unknown) {
+		switch (eventName) {
+			case "KeyGenCommitted": {
+				// A participant has committed to the new key gen
+				// Ignore if not in "collecting_commitments" state
+				if (this.#keyGenState.id !== "collecting_commitments") return;
+				// Parse event from raw data
+				const event = keyGenCommittedEventSchema.parse(eventArgs);
+				// Verify that the group corresponds to the next epoch
+				if (this.#keyGenState.groupId !== event.gid) return;
+				// This will be handled by the keyGenClient
+				await this.#keyGenClient.handleKeygenCommitment(
+					event.gid,
+					event.identifier,
+					event.commitment.c.map((c) => toPoint(c)),
+					{
+						r: toPoint(event.commitment.r),
+						mu: event.commitment.mu,
+					},
+				);
+				// If all participants have committed update state to "collecting_shares"
+				if (event.committed) {
+					const nextEpoch = this.#keyGenState.nextEpoch;
+					this.#keyGenState = {
+						id: "collecting_shares",
+						groupId: event.gid,
+						nextEpoch,
+					};
+				}
+				return;
+			}
+			case "KeyGenSecretShared": {
+				// A participant has submitted secret share for new group
+				// Ignore if not in "collecting_shares" state
+				if (this.#keyGenState.id !== "collecting_shares") return;
```

Comment:
This should be an error if `this.#keyGenState.groupId !== event.gid` IMO.

Reply:
currently we listen to all events, and the coordinator contract can be used by others to do keygen too, so I would just ignore it right now